subsys: jwt: add PSA crypto alternative to legacy libraries (TinyCrypt for ECDSA and Mbed TLS for RSA)

valeriosetti opened the pull request:
Following the discussion of the Security WG meeting of September, 9th, TinyCrypt should be deprecated soon and `jwt` is the only place in Zephyr's codebase where there is no PSA/Mbed TLS alternative to TinyCrypt to perform ECDSA.
The main goal of this PR is to solve this limitation, but it also add some extra improvements:

- a little reshape of `subsys/jwt` code, isolating signature functions;
- improving the test coverage;
- fix test's `conf` file.

valeriosetti committed:
jwt: reshape and add alternative for ECDSA using PSA

This commit:

- creates 2 new files, jwt_ecdsa.c and jwt_rsa.c, to hold the
  implementations of the corresponding ECDSA/RSA signature
  algorithms;
- RSA signature is stil done through Mbed TLS's PK module which
  can optionally make use of PSA (if enabled);
- ECDSA signature will instead use PSA, if possible, or TinyCrypt
  as fallback.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
test: subsys: jwt: fix testing of the features

- Some files (PEM/DER files, jwt-test-cert.c, user-tls-conf.h)
  were removed because they are not used or no longer necessary.
  Private keys used in the test are now in the jwt-test-private.c
  file as arrays.
- testcase.yaml has been improved in order to test all possible
  use cases: ECDSA with TC, ECDSA with PSA, RSA;
- unnecessary Kconfigs were removed from prj.conf.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

d3zd3z commented on subsys/jwt/Kconfig:
```diff
@@ -16,19 +16,45 @@ choice
+
+if JWT_SIGN_ECDSA
+
+config JWT_SIGN_ECDSA_PSA
```
Just curious, does the PSA configuration basically assume/assert that there is a proper CSPRNG enabled? Otherwise, should we check that. Signing multiple messages with the same nonce in EDCSA is a devastating failure.

valeriosetti replied:
I would say that the answer is yes, but it might depend on your configuration/platform capabilities.
I've recently done some change on how the Mbed TLS's PSA can handle random generation. You have basically [2 choices](https://github.com/zephyrproject-rtos/zephyr/blob/ccc4357f0a2c4e1b2cc39f48c227f3dc905ba3d0/modules/mbedtls/Kconfig.tls-generic#L464):

- _legacy_, where `psa_generate_random()` calls to the legacy Mbed TLS modules (i.e. entropy +  ctr/hmac drbg) to get random values; in this case the entropy module [calls back ](https://github.com/zephyrproject-rtos/zephyr/blob/ccc4357f0a2c4e1b2cc39f48c227f3dc905ba3d0/modules/mbedtls/zephyr_entropy.c#L34)to Zephyr to poll some random values. It first tries with CS sources, then it falls back to non-CS ones.

- _external_, where `psa_generate_random()` calls to CS-only sources. In this case there is the possibility to use non-CS sources (via `CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG`), but there are warning during the build that clearly state that this is meant to be used only as test.

d3zd3z reviewed (commented):
Looks good. Probably worth going through the clang-format suggestions just to keep the warnings down. Most seem to have to do with line breaks in function argument lists.

valeriosetti commented:
> Probably worth going through the clang-format suggestions just to keep the warnings down

I should have addressed most of them. There should be only complains about adding a space after casts, ex: `(void) variable` instead of `(void)variable`. Current clang-format wants the latter , but I prefer the former version. Wdyt?

In case my preference makes sense, can I edit the current `.clang-format` file adding `SpaceAfterCStyleCast: True`?

tomi-font reviewed (commented):
Looks good. Just some minor comments.

valeriosetti commented on subsys/jwt/Kconfig:
```diff
@@ -16,19 +16,45 @@ choice
-endchoice
+endif # JWT_SIGN_ECDSA
+
+if JWT_SIGN_RSA
+
+config JWT_SIGN_RSA_PSA
```
In theory yes, but with the current status the Kconfigs to select PSA/TC/MbedTLS are hidden, whereas changing it to a `choice` make it available to the user. That's an option anyway, so I'll give it a try

valeriosetti commented on subsys/jwt/Kconfig:
```diff
@@ -16,19 +16,45 @@ choice
+	bool
+	default y if !PSA_CRYPTO_CLIENT
+	depends on CSPRNG_ENABLED
+	select MBEDTLS_KEY_EXCHANGE_RSA_ENABLED
```
Yes,  indeed, but changing this to a `choice` as you suggested [above](https://github.com/zephyrproject-rtos/zephyr/pull/78243#discussion_r1754274614) solved the problem

valeriosetti commented on subsys/jwt/jwt_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+
+	return 0;
+}
+#endif /* CONFIG_JWT_SIGN_ECDSA_TC */
+
+#if defined(CONFIG_JWT_SIGN_ECDSA_PSA)
```
Yes, but this has been moved to another file due to a refactoring, so this does not hold anymore.

valeriosetti commented on subsys/jwt/jwt_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+	size_t sig_len_out;
+	int ret;
+
+	psa_set_key_type(&attr, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
```
I'm pretty sure the answer is yes, but I will check

valeriosetti commented on subsys/jwt/Kconfig:
```diff
@@ -16,19 +16,45 @@ choice
+	default y if PSA_CRYPTO_CLIENT
+	select PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT
+	select PSA_WANT_ALG_ECDSA
+	select PSA_WANT_ECC_SECP_R1_256
```
Nice catch! Thanks!

valeriosetti commented on subsys/jwt/jwt_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+	}
+
+exit:
+	psa_reset_key_attributes(&attr);
```
This [might free additional resources](https://github.com/Mbed-TLS/mbedtls/blob/8ea8b30941a6d364f5bf4b3850f5312112e760d9/tf-psa-crypto/include/psa/crypto.h#L391) allocated with the attributes, but after checking the current implementation in Mbed TLS, there's no extra memory allocated with `psa_key_attributes_t` so `psa_reset_key_attributes` is not really useful indeed

tomi-font replied:
Yeah. Well, because of how it's documented, then any way is fine by me. I imagine the unnecessary `psa_reset_key_attributes()` would be optimized out by the compiler if it does nothing interesting.

valeriosetti commented on subsys/jwt/jwt_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+	status = psa_import_key(&attr, der_key, der_key_len, &key_id);
+	if (status != PSA_SUCCESS) {
+		ret = -EINVAL;
+		goto exit;
```
Now that I removed the `psa_reset_key_attributes` yes (previously not)

valeriosetti commented on subsys/jwt/jwt_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+	return res == TC_CRYPTO_SUCCESS ? 0 : -EINVAL;
+}
+
+int default_CSPRNG(uint8_t *dest, unsigned int size)
```
At first I thought yes, but changing this I realized that this is not possible. I'll add a comment on top of this function to explain why

tomi-font commented on subsys/jwt/Kconfig:
```diff
@@ -7,27 +8,74 @@ menuconfig JWT
+	prompt "Select crypto library to be used"
+
+config JWT_USE_PSA
+	bool "PSA crypto API library"
```
Don't you want to add some `depends on` here?

valeriosetti replied:
Right. I implemented what you proposed [here](https://github.com/zephyrproject-rtos/zephyr/pull/78243#discussion_r1758327347)

tomi-font replied:
As this is changing the default behavior (now defaulting to PSA, enabling what's needed) we may want a migration guide entry.

tomi-font commented on subsys/jwt/Kconfig:
```diff
@@ -7,27 +8,74 @@ menuconfig JWT
+
+# Hidden Kconfigs
+
+config JWT_SIGN_ECDSA_PSA
+	bool
+	default y
+	depends on JWT_SIGN_ECDSA && JWT_USE_PSA
+	select MBEDTLS if !BUILD_WITH_TFM
+	select MBEDTLS_PSA_CRYPTO_C if !BUILD_WITH_TFM
+	select PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT
+	select PSA_WANT_ALG_ECDSA
+	select PSA_WANT_ECC_SECP_R1_256
+	select PSA_WANT_ALG_SHA_256
+
+config JWT_SIGN_ECDSA_LEGACY
```
Why not a choice here as well? Would make sense to me. Especially since you are using `default y` for both configs, with incompatible conditions. Just introduce a `choice`, remove those `default y`, keeping the `depends on` conditions. No need to even put a `default` in the `choice`.

tomi-font commented on subsys/jwt/Kconfig:
```diff
@@ -7,27 +8,74 @@ menuconfig JWT
+	bool
+	default y
+	depends on JWT_SIGN_ECDSA && JWT_USE_PSA
+	select MBEDTLS if !BUILD_WITH_TFM
+	select MBEDTLS_PSA_CRYPTO_C if !BUILD_WITH_TFM
```
Suggest moving those under `JWT_USE_PSA` for this and `RSA_PSA` to avoid repetition.

tomi-font commented on subsys/jwt/jwt.c:
```diff
@@ -190,120 +180,27 @@ int jwt_add_payload(struct jwt_builder *builder,
-	if (res != 0) {
-		return res;
-	}
-	uECC_set_rng(&default_CSPRNG);
```
This call was not needed?

valeriosetti replied:
Yeah, it's a bit tricky. `uECC_set_rng` allows you to define a custom RNG function, but if you check in `modules/crypto/tinycrypt/lib/source/ecc.c` you'll see that:
```
#if default_RNG_defined
static uECC_RNG_Function g_rng_function = &default_CSPRNG;
#else
static uECC_RNG_Function g_rng_function = 0;
#endif
```
and `default_RNG_defined` is always set (defined in `modules/crypto/tinycrypt/lib/include/tinycrypt/ecc_platform_specific.h`).
So that line that I removed just sets the very same function that is already set by default in TinyCrypt. Not really useful

tomi-font commented on subsys/jwt/CMakeLists.txt:
```diff
@@ -2,4 +2,9 @@
 zephyr_library()
 zephyr_library_sources(jwt.c)
+
+zephyr_library_sources_ifdef(CONFIG_JWT_SIGN_ECDSA_LEGACY jwt_ecdsa.c)
+zephyr_library_sources_ifdef(CONFIG_JWT_SIGN_RSA_LEGACY jwt_rsa.c)
```
IMO would make more sense to name those files something `legacy` if they're not holding the ECDSA/RSA implementation in all cases (in the PSA case it's in `jwt_psa.c`.

tomi-font commented on subsys/jwt/jwt.c:
```diff
@@ -190,120 +180,27 @@ int jwt_add_payload(struct jwt_builder *builder,
-	int res = tc_ctr_prng_generate(&prng_state, NULL, 0, dest, size);
-	return res;
-}
+int jwt_sign_impl(struct jwt_builder *builder, const unsigned char *der_key,
+		  size_t der_key_len, unsigned char *sig, size_t sig_size);
```
Not introduced in this PR, but it would be better to have this function declared in a header that's then included in the relevant places...

tomi-font commented on subsys/jwt/jwt_psa.c:
```diff
@@ -0,0 +1,50 @@
+{
+	psa_status_t status;
+	psa_key_attributes_t attr = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = 0;
```
Minor, do as you wish: ` = 0` not needed.

tomi-font commented on subsys/jwt/jwt_psa.c:
```diff
@@ -0,0 +1,50 @@
+	psa_set_key_algorithm(&attr, PSA_ALG_RSA_PKCS1V15_SIGN(PSA_ALG_SHA_256));
+	alg = PSA_ALG_RSA_PKCS1V15_SIGN(PSA_ALG_SHA_256);
+#endif /* CONFIG_JWT_SIGN_ECDSA || CONFIG_JWT_SIGN_RSA */
+	psa_set_key_usage_flags(&attr, PSA_KEY_USAGE_SIGN_MESSAGE);
```
Used to be `PSA_KEY_USAGE_SIGN_HASH` for ECDSA?

valeriosetti replied:
I was wrong before, sorry. What we want to do here is hash and then sign, so `psa_sign_message()` is the function we are [looking for](https://github.com/zephyrproject-rtos/mbedtls/blob/fb36f3fe20f9f62e67b1a20c0cfe0a6788ec2bf6/include/psa/crypto.h#L2832). The previous `psa_sign_hash()` [assumes](https://github.com/zephyrproject-rtos/mbedtls/blob/fb36f3fe20f9f62e67b1a20c0cfe0a6788ec2bf6/include/psa/crypto.h#L2951) that you manually compute the hash before computing the signature

tomi-font commented on subsys/jwt/jwt_psa.c:
```diff
@@ -0,0 +1,50 @@
+#endif /* CONFIG_JWT_SIGN_ECDSA || CONFIG_JWT_SIGN_RSA */
+	psa_set_key_usage_flags(&attr, PSA_KEY_USAGE_SIGN_MESSAGE);
+
+	status = psa_import_key(&attr, der_key, der_key_len, &key_id);
```
No more setting of the key bits?

valeriosetti replied:
Yes, it's not strictly needed. The key data being passed will automatically determine the key's length: https://github.com/zephyrproject-rtos/mbedtls/blob/fb36f3fe20f9f62e67b1a20c0cfe0a6788ec2bf6/include/psa/crypto.h#L586

tomi-font commented on subsys/jwt/Kconfig:
```diff
@@ -7,27 +8,78 @@ menuconfig JWT
+# the JWT. User's selections on the above choices will determine which element
+# will be picked here.
+choice
+	prompt "Define which signature algorithm and crypto library is used to sign the JWT"
```
I think that you want to just remove this `prompt` as well as those of the Kconfig options inside of this `choice`. They're not meant to have prompts, but to be automatically determined by the selections above, right?

valeriosetti replied:
That's what I did initially, but then there are failures like:
`warning: the choice symbol JWT_SIGN_ECDSA_PSA (defined at subsys/jwt/Kconfig:47) has no prompt`

I agree that having a prompt makes little sense for these derived Kconfigs. I think it's better to go back to the previous implementation where these were not part of `choice`. Wdyt?

tomi-font replied:
Ah, damn. Well, whichever way then. In a sense you could even get completely rid of those extra Kconfig options and do all the logic in the PSA/LEGACY or RSA/ECDSA ones. But that would required having every `select` conditioned by an `if`, which is not great.

valeriosetti committed:
doc: update JWT documentation

Update JWT subsystem documentation concerning the changes related
to:

- default library used
- new Kconfigs added

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -288,5 +288,10 @@ Shell
+JWT (JSON Web Token)
+====================
+
+* Default signature library is now PSA Crypto APIs instead of Mbed TLS. (:github:`78243`)
```
I think this could be a little bit more explanatory. The other option is not even necessarily Mbed TLS, it's either Mbed TLS or TinyCrypt depending on the signature algorithm. It could be nice also to mention the relevant Kconfig options that the user can enable/disable to have an influence on this.

valeriosetti replied:
I kept it short (maybe too much) because I didn't want to duplicate the content of the release-notes. For a migration-guide I thought that simply listing the default behavior changes should be enough. After all the Konfigs to select ECDSA/RSA did not changed in this PR and RSA is still the default. What I added here is the possibility to choose between legacy or PSA APIs (but these new Kconfigs are already in the release-notes) and this is where the change happened.
So IMO a possible improvement of this comment is that PSA APIs are the default library that will be used for both RSA/ECDSA. New Kconfigs to switch back to legacy libraries are in the release-notes.
How does this sounds?

tomi-font replied:
RSA is still the default, but the previous (default) behavior of using Mbed TLS/TC is changed (to use PSA). I think it's good to mention enough information in the migration guide itself. You could simply add that to stick to the old (deprecated) behavior, explicitly enable `CONFIG_JWT_USE_LEGACY`.

tomi-font commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -361,6 +361,14 @@ Libraries / Subsystems
 
+* JWT (JSON Web Token)
+
+  * New Kconfigs were added to specify which library to use for the signature:
+
+    * :kconfig:option:`CONFIG_JWT_USE_PSA` (default) use PSA Crypto APIs;
```
```suggestion
  * The following new Kconfigs were added to specify which library to use for the signature:

    * :kconfig:option:`CONFIG_JWT_USE_PSA` (default) use the PSA Crypto API;
```

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -288,5 +288,12 @@ Shell
+JWT (JSON Web Token)
+====================
+
+* Default signature library is now PSA Crypto API for both RSA and ECDSA. The
+  newly added :kconfig:option:`CONFIG_JWT_USE_LEGACY` can be used to switch
```
```suggestion
* The default signature library is now PSA Crypto API for both RSA and ECDSA. The
  newly-added :kconfig:option:`CONFIG_JWT_USE_LEGACY` can be used to switch
```
Also, I don't know if we can call the PSA Crypto API a library. The library is what implements the API, not the API itself. But maybe I'm being too pedantic here... Do what you want. :slightly_smiling_face:

dkalowsk commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -288,5 +288,12 @@ Shell
+
+* By default signature is now computed through PSA Crypto API for both RSA and ECDSA.
+  The newly-added :kconfig:option:`CONFIG_JWT_USE_LEGACY` can be used to switch
+  back to previous libraries (TinyCrypt for ECDSA and Mbed TLS for RSA). (:github:`78243`)
```
Would it make sense to add a note here along the lines of:
```
The conversion to the PSA Crypto API is being done in preparation for the deprecation of TinyCrypt.
```
Not everyone follows the security working groups minutes (nor do they read migration guides).  The thought was to to make it more clear the deprecation is happening in multiple areas.

tomi-font commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -288,5 +288,14 @@ Shell
+  The newly-added :kconfig:option:`CONFIG_JWT_USE_LEGACY` can be used to switch
+  back to previous libraries (TinyCrypt for ECDSA and Mbed TLS for RSA).
+  The conversion to the PSA Crypto API is being done in preparation for the
+  deprecation of TinyCrypt.(:github:`78243`)
```
```suggestion
  deprecation of TinyCrypt. (:github:`78243`)
```

\+ could add a link to https://github.com/zephyrproject-rtos/zephyr/issues/43712?

dkalowsk commented on subsys/jwt/jwt.h:
```diff
@@ -0,0 +1,8 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+int jwt_sign_impl(struct jwt_builder *builder, const unsigned char *der_key,
```
you may want some header guards here.  Additionally for this to be stand-alone the include where `jwt_builder` is defined should be added to the file... or forward declared as it's a pointer to force the compiler to warn about transitive includes.

valeriosetti replied:
You're absolutely right. Thanks for both suggestions!

dkalowsk commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -288,5 +288,14 @@ Shell
+JWT (JSON Web Token)
+====================
+
+* By default signature is now computed through PSA Crypto API for both RSA and ECDSA.
```
I think you're missing a word here... `By default, <something> signature is now... `

valeriosetti replied:
Rather than a word I would say an article, "the". JWT assumes that the JSON content is signed and "signature" is the subject of this sentence, so I would say that "by default, the signature" should be fine.
However I'm not a native english speaker, so I might miss something so please I apologize for any mistake.

dkalowsk reviewed (commented):
Minor comment, with the rest looking overall good to me.

d3zd3z commented:
> > Probably worth going through the clang-format suggestions just to keep the warnings down
> 
> I should have addressed most of them. There should be only complains about adding a space after casts, ex: `(void) variable` instead of `(void)variable`. Current clang-format wants the latter , but I prefer the former version. Wdyt?
> 
> In case my preference makes sense, can I edit the current `.clang-format` file adding `SpaceAfterCStyleCast: True`?

Not using a space between a cast and the value is almost universal across programming conventions. I'd suggest getting used to it.

d3zd3z commented on subsys/jwt/jwt_legacy_ecdsa.c:
```diff
@@ -0,0 +1,81 @@
+	sys_rand_get(entropy, sizeof(entropy));
+
+	int res = tc_ctr_prng_init(&prng_state, (const uint8_t *) &entropy, sizeof(entropy),
+				   personalize, sizeof(personalize));
```
I'd suggest fixing this, not using a space after a cast is near universal in C style conventions.

d3zd3z commented on subsys/jwt/jwt_legacy_ecdsa.c:
```diff
@@ -0,0 +1,81 @@
+	struct tc_sha256_state_struct ctx;
+	uint8_t hash[32];
+	int res;
+	(void) sig_size;
```
Please use `ARG_UNUSED(sig_size);`

d3zd3z commented on subsys/jwt/jwt_psa.c:
```diff
@@ -0,0 +1,51 @@
+	psa_set_key_type(&attr, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_algorithm(&attr, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	alg = PSA_ALG_ECDSA(PSA_ALG_SHA_256);
+#else /* CONFIG_JWT_SIGN_RSA */
```
This is rather silly and picky of it, but might as well comply.

d3zd3z commented on subsys/jwt/jwt_psa.c:
```diff
@@ -0,0 +1,51 @@
+		return -EINVAL;
+	}
+
+	status = psa_sign_message(key_id, alg,
```
I care less about this, since I think your formatting is slightly more readable, as it groups the things that go together.

d3zd3z commented on tests/subsys/jwt/src/jwt-test-private.c:
```diff
@@ -4,109 +4,130 @@
+ * - get the first OCTECT STRING, which is the private key, and copy it here
+ */
 
 unsigned char jwt_test_private_der[] = {
```
In this case, the clang format output is actively terrible. If you have ideas on improving the config for this one, perhaps that should be proposed. But, not in this PR, just leave it.

d3zd3z reviewed (changes requested):
I think you should fix the spacing on casts, and use the `ARG_UNUSED()` macro. The others from clang-format are not good, and maybe we want to propose fixing that in another PR.
